pyomo/contrib/pynumero/algorithms/solvers/cyipopt_solver.py

```diff
@@ -9,61 +9,58 @@
-objects for the matrices (e.g., AmplNLP and PyomoNLP)
+The cyipopt_solver module includes two solvers that call CyIpopt. One,
+CyIpoptSolver, is a solver that operates on a CyIpoptProblemInterface
+(such as CyIpoptNLP). The other, PyomoCyIpoptSolver, operates directly on a
+Pyomo model.
+
 """
 import io
 import sys
 import logging
 import os
 import abc
 
+from pyomo.common.deprecation import relocated_module_attribute
 from pyomo.common.dependencies import (
     attempt_import,
     numpy as np, numpy_available,
 )
 from pyomo.common.tee import redirect_fd, TeeStream
 
-def _cyipopt_importer():
-    import cyipopt
-    # cyipopt before version 1.0.3 called the problem class "Problem"
-    if not hasattr(cyipopt, 'Problem'):
-        cyipopt.Problem = cyipopt.problem
-    # cyipopt before version 1.0.3 put the __version__ flag in the ipopt
-    # module (which was deprecated starting in 1.0.3)
-    if not hasattr(cyipopt, '__version__'):
-        import ipopt
-        cyipopt.__version__ = ipopt.__version__
-    # Beginning in 1.0.3, STATUS_MESSAGES is in a separate
-    # ipopt_wrapper module
-    if not hasattr(cyipopt, 'STATUS_MESSAGES'):
-        import ipopt_wrapper
-        cyipopt.STATUS_MESSAGES = ipopt_wrapper.STATUS_MESSAGES
-    return cyipopt
-
-cyipopt, cyipopt_available = attempt_import(
-     'ipopt',
-     error_message='cyipopt solver relies on the ipopt module from cyipopt. '
-     'See https://github.com/mechmotum/cyipopt.git for cyipopt '
-     'installation instructions.',
-     importer=_cyipopt_importer,
-)
-
 # Because pynumero.interfaces requires numpy, we will leverage deferred
 # imports here so that the solver can be registered even when numpy is
 # not available.
 pyomo_nlp = attempt_import('pyomo.contrib.pynumero.interfaces.pyomo_nlp')[0]
 pyomo_grey_box = attempt_import('pyomo.contrib.pynumero.interfaces.pyomo_grey_box_nlp')[0]
 egb = attempt_import('pyomo.contrib.pynumero.interfaces.external_grey_box')[0]
 
+# Defer this import so that importing this module (PyomoCyIpoptSolver in
+# particular) does not rely on an attempted cyipopt import.
+cyipopt_interface, _ = attempt_import(
+    'pyomo.contrib.pynumero.interfaces.cyipopt_interface'
+)
+
+# These attributes should no longer be imported from this module. These
+# deprecation paths provide a deferred import to these attributes so (a) they
+# can still be used until these paths are removed, and (b) the imports are not
+# triggered when this module is imported.
+relocated_module_attribute(
+    "cyipopt_available",
+    "pyomo.contrib.pynumero.interfaces.cyipopt_interface.cyipopt_available",
+    "6.5.1.dev0",
+)
+relocated_module_attribute(
+    "CyIpoptProblemInterface",
+    "pyomo.contrib.pynumero.interfaces.cyipopt_interface.CyIpoptProblemInterface",
+    "6.5.1.dev0",
+)
+relocated_module_attribute(
+    "CyIpoptNLP",
+    "pyomo.contrib.pynumero.interfaces.cyipopt_interface.CyIpoptNLP",
+    "6.5.1.dev0",
+)
```

Comment:
I commonly import `cyipopt_available` from this module, so I added these relocation warnings so these imports will still be available. This seems not to increase import time, judging by test_environ.py still passing locally.